### PR TITLE
fix: Various NPEs on terminated managed ledger

### PR DESCRIPTION
**Pull request by dlg99**
*(If this PR fixes a github issue, please add `Fixes #<xyz>`.)*

### Motivation

I encountered this while experimenting with repros for issue: #12070 

The problem here is that various methods on terminated managed ledger throw NPEs after broker restart. 
The rootcause is that `currentLedger` field is not expected to be set in case of terminated ledger hence stays null; the null check on that field is inconsistent.

### Modifications

Add  null check everywhere the `currentLedger` is accessed, added unit test.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change added unit tests

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

No

### Documentation

Need to update docs? 

- [X] no-need-doc

### CODE REVIEW

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ConsumedLedgersTrimTest.java [hunk 1]
```diff
@@ -170,8 +171,89 @@ public void testConsumedLedgersTrimNoSubscriptions() throws Exception {
...
+        // set retention parameters, the ledgers are to be deleted as soon as possible
+        // but the topic is not to be automatically deleted
+        PersistentTopic persistentTopic = (PersistentTopic) pulsar.getBrokerService().getOrCreateTopic(topicName).get();
+        ManagedLedgerConfig managedLedgerConfig = persistentTopic.getManagedLedger().getConfig();
+        managedLedgerConfig.setRetentionSizeInMB(-1);
+        managedLedgerConfig.setRetentionTime(100000, TimeUnit.SECONDS);
+        managedLedgerConfig.setMaxEntriesPerLedger(2);
+        managedLedgerConfig.setMinimumRolloverTime(1, TimeUnit.MILLISECONDS);
+        MessageId initialMessageId = persistentTopic.getLastMessageId().get();
+        log.info("lastmessageid " + initialMessageId);
```
Comment (eolivelli) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ConsumedLedgersTrimTest.java:205]: Nit: use parameters

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java [hunk 2]
```diff
@@ -1602,7 +1605,7 @@ synchronized void createLedgerAfterClosed() {
     @Override
     public void rollCurrentLedgerIfFull() {
         log.info("[{}] Start checking if current ledger is full", name);
-        if (currentLedgerEntries > 0 && currentLedgerIsFull()) {
+        if (currentLedger != null && currentLedgerEntries > 0 && currentLedgerIsFull()) {
```
Comment (eolivelli) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:1608]: Will it make sense to exit early from this method if there is no 'currentLedger' ?

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java [hunk 3]
```diff
@@ -770,13 +770,15 @@ private synchronized void internalAsyncAddEntry(OpAddEntry addOperation) {
             currentLedgerSize += addOperation.data.readableBytes();
 
             if (log.isDebugEnabled()) {
-                log.debug("[{}] Write into current ledger lh={} entries={}", name, currentLedger.getId(),
+                log.debug("[{}] Write into current ledger lh={} entries={}", name,
+                        currentLedger == null ? null : currentLedger.getId(),
```
Comment (eolivelli) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:774]: For these logging cases we could add a Long  getCurrentLedgerId() method that returns null is there is no currentLedger.
We will save some code duplication

**Review by eolivelli — CHANGES_REQUESTED** [review]
So basically here we are trying to not fail in case currentLedger is null.
And you found a case in which currentLedger is null.

The patch makes sense to me.

@codelipenghui @merlimat @rdhabalia cN you please take a look?

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java (hunk 2 above)
Comment (dlg99) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:1608]: essentially this is it.

**Comment by sijie** [comment]
@codelipenghui @merlimat Can you take a look at this PR?

**Review by eolivelli — APPROVED** [review]

**Review by lhotari — APPROVED** [review]

**Review by michaeljmarshall — APPROVED** [review]

**Comment by tisonkun** [comment]
Closed as stale and conflict. Please rebase and resubmit the patch if it's still relevant.

This one should be good to go. But often, a stale PR over a year has little chance of getting updated again. Rebase and resubmit is always available and welcome :)

**Pull request closed without merging**